### PR TITLE
fix: Correctly persist createdAt attribute

### PR DESCRIPTION
**Pull request by jakubkoci**
I needed to move `createAt` attribute initialization up in the inheritance chain to every particular record. I'm not sure if this is a good approach in the long-term but it at least fixes the issue for now. I also aligned `id` attribute initialization across all records, so even if this is a wrong approach we have it wrong in the same way everywhere :)

### CODE REVIEW

File: src/lib/storage/BasicMessageRecord.ts [hunk 1]
```diff
@@ -17,7 +18,7 @@ export class BasicMessageRecord extends BaseRecord implements BasicMessageStorag
   public readonly type = BasicMessageRecord.type;
 
   public constructor(props: BasicMessageStorageProps) {
-    super(props.id || uuid());
+    super(props.id || uuid(), props.createdAt || Date.now());
```
Comment (TimoGlastra) [inline src/lib/storage/BasicMessageRecord.ts:21]: TypeScript 4.0 introduced some new [short circuit assignment operators](https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/#short-circuiting-assignment-operators). Which are perfect for this usage.

```suggestion
    super(props.id ?? uuid(), props.createdAt ?? Date.now());
```

File: src/lib/storage/ConnectionRecord.ts [hunk 2]
```diff
@@ -49,7 +51,7 @@ export class ConnectionRecord extends BaseRecord implements ConnectionStoragePro
   public readonly type = ConnectionRecord.type;
 
   public constructor(props: ConnectionStorageProps) {
-    super(props.id);
+    super(props.id || uuid(), props.createdAt || Date.now());
```
Comment (TimoGlastra) [inline src/lib/storage/ConnectionRecord.ts:54]: Same

File: src/lib/storage/CredentialRecord.ts [hunk 3]
```diff
@@ -27,7 +28,7 @@ export class CredentialRecord extends BaseRecord implements CredentialStoragePro
   public state: CredentialState;
 
   public constructor(props: CredentialStorageProps) {
-    super(props.id ? props.id : uuid());
+    super(props.id || uuid(), props.createdAt || Date.now());
```
Comment (TimoGlastra) [inline src/lib/storage/CredentialRecord.ts:31]: Same

File: src/lib/storage/IndyStorageService.test.ts [hunk 4]
```diff
@@ -18,20 +19,20 @@ class TestRecord extends BaseRecord {
   public readonly type = TestRecord.type;
 
   public constructor(props: TestRecordProps) {
-    super(props.id || uuid());
+    super(props.id || uuid(), props.createdAt || Date.now());
```
Comment (TimoGlastra) [inline src/lib/storage/IndyStorageService.test.ts:22]: Same

File: src/lib/storage/ProvisioningRecord.ts [hunk 5]
```diff
@@ -15,7 +17,7 @@ export class ProvisioningRecord extends BaseRecord {
   public readonly type = ProvisioningRecord.type;
 
   public constructor(props: ProvisioningRecordProps) {
-    super(props.id);
+    super(props.id || uuid(), props.createdAt || Date.now());
```
Comment (TimoGlastra) [inline src/lib/storage/ProvisioningRecord.ts:20]: Same

**Review by TimoGlastra — APPROVED** [review]
Looks good, just the suggestion to use the new short circuit operator

**Comment by jakubkoci** [comment]
> Looks good, just the suggestion to use the new short circuit operator

Hmm, that interesting, but how does it actually work? I'm looking around but can't find anything about that 😄

**Comment by TimoGlastra** [comment]
> > Looks good, just the suggestion to use the new short circuit operator
> 
> Hmm, that interesting, but how does it actually work? I'm looking around but can't find anything about that 😄

I linked to the wrong thing. What I actually mean is ["nullish coalescing"](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#nullish-coalescing)

Basically when you use `||` it will look if the left hand side results in a truthy or falsy value. This means if I provide `createdAt`. value of 0 it will equal to falsy and use `Date.now`. With `??` it will check whether the variable has a value (so not `null` or `undefined`). This means that a value of 0 will not be seen as falsy and will be used.

File: src/lib/storage/BasicMessageRecord.ts (hunk 1 above)
Comment (jakubkoci) [reply src/lib/storage/BasicMessageRecord.ts:21]: Ah, I see. That's nice. I can't even count how often I solve issues with `if (number) ...` 😄

Comment (jakubkoci) [reply src/lib/storage/BasicMessageRecord.ts:21]: Hmm, as I'm thinking about it. Are they really perfect for this usage? Maybe values `''` and `0` are not quite valid as `id` or `createAt`

Comment (jakubkoci) [reply src/lib/storage/BasicMessageRecord.ts:21]: But that's perhaps the responsibility of some eventual validation. Ok, I'll change it to `??`

Comment (TimoGlastra) [reply src/lib/storage/BasicMessageRecord.ts]: Well yeah, 0 is probably not a commonly used value. However it is unexpected behaviour IMO to change the value without notic. When I pass a value of `0` and it is automatically changed to `Date.now` without warning or error that is weird I think. Same with empty string.

**Pull request merged**